js/node-manager/src/client/agents/node-manager.tsx

```diff
@@ -0,0 +1,426 @@
+  immutable: ImmutableContainer<T>,
+  fn: (t: Readonly<T>) => R,
+): (() => R) => {
+  let lastContainerState = immutable.access()
+  let cache = ImmutableLazyCalcUninit as typeof ImmutableLazyCalcUninit | R
+
+  return () => {
+    if (immutable.access() !== lastContainerState) {
+      lastContainerState = immutable.access()
+      cache = fn(immutable.access())
+    }
+    const ret = cache !== ImmutableLazyCalcUninit ? cache : (fn(immutable.access()) as R)
+    cache = ret
+    return ret
+  }
+}
+
+const intoUiNode = (
+  addr: string,
+  connection: ConnectionResult | null,
+  nodeInfo: GetNodeDetailsResponse | null,
+): UiNode => {
+  if (!connection) return { type: NodeType.Connecting, addr, prevError: null }
+
+  if (E.isLeft(connection)) return { type: NodeType.Connecting, addr, prevError: connection.left }
+  const { peer } = connection.right
+
+  if (!nodeInfo) return { type: NodeType.Connected, addr, peer }
+  if (nodeInfo.type === NodeType.Reachable) return { ...nodeInfo, addr, timeouts: 0 }
+  return { ...nodeInfo, addr }
+}
+
+type Internals = {
+  /**
+   * User-indicated addresses that needs to be tracked
+   * IMMUTABLE, must be comparable with nodes === nodes
+   */
+  trackedAddresses: ImmutableContainer<Set<string>>
+  /**
+   * "Connection" to nodes, containing either peer info or error
+   * IMMUTABLE, must be comparable with nodes === nodes
+   */
+  connections: ImmutableContainer<Map<string, ConnectionResult>>
+  /**
+   * Contains agents providing node details asynchonously
+   */
+  nodeDetailsProviders: ImmutableContainer<Map<string, NodeInfoProvider>>
+  offsets: null | OffsetInfo
+}
+
+type ConnectionResult = E.Either<string, { peer: string }>
+
+const getDisconnectibles = (data: Internals) => {
+  const connections = data.connections.access()
+  const addresses = data.connections.access()
+  return Array.from(connections.keys()).filter((addr) => !addresses.has(addr))
+}
+
+const getRemoveableDetailsProviders = (data: Internals) => {
+  const connections = data.connections.access()
+  const providers = data.nodeDetailsProviders.access()
+  return Array.from(providers.keys()).filter((addr) => !connections.has(addr))
+}
+
+/**
+ * Remove nodes whose addresses are untracked.
+ */
+const regulateRemovals = (data: Internals, onChange: () => unknown) => {
+  const disconnectibles = getDisconnectibles(data)
+  const removableNodeDetailsProviders = getRemoveableDetailsProviders(data)
+
+  if (disconnectibles.length > 0) {
+    data.connections.mutate((nodes) => {
+      disconnectibles.forEach((addr) => {
+        nodes.delete(addr)
+
+        console.log('disconnect from:', addr)
+      })
+    })
+  }
+
+  if (removableNodeDetailsProviders.length > 0) {
+    data.nodeDetailsProviders.mutate((nodeInfoProviders) => {
+      removableNodeDetailsProviders.forEach((addr) => {
+        nodeInfoProviders.get(addr)?.destroy()
+        nodeInfoProviders.delete(addr)
```

Comment:
Well they might want to avoid extra allocation because optimizing stuff (such as not caring about the returned value) in an interpreted language might introduce resource contest with the execution of the code.
Or maybe it was the case of "move fast" and not realizing that "break things" isn't an option for such a prevalent product.